blog:bot helper CI Job (#293)
kubernetes/contributor-site: content/en/blog/2022/2022-03-15-k8s-triage-bot-helper-ci-job/K8s Triage bot Helper CI Job_ ‘make update’.md

## Why is this needed?
 
While contributing to [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes) for some minor documentation changes, the pushed changes needed to be updated with other verified contents of the entire documentation. So, in order for the change to take effect, a single command must be performed to ensure that all tests on the CI pipeline pass. The single command `make update` runs all presubmission verification tests. For some reason on the windows `wsl` environment the tests, specifically the [update-openapi-spec.sh](https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-openapi-spec.sh) script, failed (in my case, take a look at the conversation [here](https://github.com/kubernetes/kubernetes/pull/107691)), eventually failing the `pull-kubernetes-verify` tests.

### System info:

| OS name | Version | System type |Memory| WSL |
| ------- | ------- | ----------- | ---  | --- |
|Windows 11 home|10.0.22000 Build 22000|x64-based|8.00 GB|Ubuntu-20.04 (v2)|

You might encounter the following on your PR

> > failing test cases notification from k8s-ci-robot

Reviewer: seems like you're missing an image here?

The tests failing the particular issue:

![Failed Test Cases](failing_test_cases.png)

Consecutively, 

![Failed Test Cases robot comment](k8s_ci_failed_tests.png)

Additionally one can check the failed test via the `link` provided under details in the above image.

### Potential solution
Run the failing `.sh` scripts individually known from the CI job output, to generate the expected files to fix up the failures. The `.sh` scripts can be found residing under the `hack/` directory at the root of the `kubernetes/kubernetes` code base. 

> > kubernetes/kubernetes &rarr; hack/ &rarr; `.sh` update files

Reviewer: ```suggestion
 `kubernetes/kubernetes &rarr; hack/ &rarr; `.sh` update files`
```
this should be a code sample, not a block quote

In this particular case, these files were to be run:

* [hack/update-generated-protobuf.sh](https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-generated-protobuf.sh) 
* [hack/update-generated-swagger-docs.sh](https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-generated-swagger-docs.sh)
* [hack/update-openapi-spec.sh](https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-openapi-spec.sh)

> ### Might run into something like these

Reviewer: ```suggestion
After you run these, you might see either of the below issues:
```

> 1. Using Codespaces

Reviewer: ```suggestion
Using Codespaces:
```

Remove digit, because it doesn't format correctly.

![codespaces error output](codespaces.png)
![codespaces error output](codespaces2.png)

> 1. Using Visual Studio Code

Reviewer: ```suggestion

Using Visual Studio Code:
```

Remove digit, add line break

![vscode error output](vscode.png)

### Potential solution to the above error:
1. Remove the Makefile and Makefile.generated_files files:

>         rm Makefile Makefile.generated_files

Reviewer: make this a code block, with explicit ``` please

2. Create the symlinks:

        ln -s build/root/Makefile Makefile

        ln -s build/root/Makefile.generated_files Makefile.generated_files

> PS: takes adequate time to generate the scripts on windows.

## The current situation

Many contributors do not have access to powerful environments in which to run `make update` or `make verify`. They can utilise *vscode/wsl/codespaces* and other tools to recommend modifications, but they might get tripped up by `make verify` scripts because in many cases, we end up with files that needs to be re-generated. It's a tall order for them to scan the build log from `make verify` to determine which specific scripts in `hack/` directory they need to run.

### Solution

Two solutions are proposed by [@dims](https://github.com/dims) for the time being:

* **Short Term:** add a CI job that folks can trigger when they need it, that runs `make update` and provides a zip archive that they can be told to download. The archive includes all the files that changed as a result of running `make update`.
*  **Long Term:** add a new Prow bot command that generates an additional commit on their PR.

> ### Short Term Solution

Reviewer: Above you have Solution, then Short Term and Long Term.  Then here you have "Short Term Solution".  How do these solutions relate to each other?

Author: The above ones are just a preview (summary) next is the procedure for short term solution. How do you suggest it to be formatted? 
like:
```
short term: preview 
        short term: explanation
long term: preview   
```

Reviewer: ok, will suggest an alternate way to arrange the text

Once we realized the problem was that some of the verify scripts needed either linux to work better or a large CPU/Memory even with linux, we decided to add a new CI job. This CI job is named “pull-kubernetes-update”. You can trigger this CI job by running “/test pull-kubernetes-update” bot command in any PR. This CI job runs `make update` and generates a zip file named “updated-files.zip” in the artifacts directory for that job. Folks can then download the zip which has the changes made when `make update` was run and update their PR with the newly updated code.

**Steps:**

> > Make sure to rebase your 'working branch' to 'master' to align with current changes made to the repository. 

Reviewer: ```suggestion
Make sure to rebase your 'working branch' to 'master' to align with current changes made to the repository. 
```

On your PR, write a new comment that contains only the text: 

    /test pull-kubernetes-update

![pull-kubernetes-update command](tpku_command.png)

You will see the automated check lists and details.

![command preview on github](checklists.png)

Once the checks are complete, click the `Details` link on `pull-kubernetes-update` job to go to the artifacts directory and  download the `updated-files.zip` file.

![automated checklist of tests appearing after update command execution](completed_checklist.png)

![completed checklist of tests](artifacts.png)

![artifacts directory](jenkins.png)

> Now, update :rocket: the PR by adding the extracted files you downloaded. 

Reviewer: ```suggestion
Now, update the PR by adding the extracted files you downloaded. 
```

Remove the emoji, which is confusing.

---
layout: blog

> title: "K8s Triage bot Helper CI Job: ‘*make update*’"

Reviewer 2: How about (using a YAML multiline string)
```suggestion
title: >-
  K8s Triage Bot Helper CI Job: "make update"
```

date: 2022-03-15
slug: k8s-triage-bot-helper-ci-job
---

**Authors:** [Subhasmita Swain](https://github.com/SubhasmitaSw), [Davanum Srinivas](https://github.com/dims) 

---

> If you are contributing to the Kubernetes project and are developing on a Windows PC, it is conceivable that you will encounter certain issues that will cause your patches to master to somehow be halted. This blog is a workaround for a similar issue I encountered when attempting to have my modifications approved and merged into the master repository.

Reviewer 2: Several factual errors here
```suggestion
If you are contributing to the Kubernetes project and are developing on a Windows PC, it is conceivable that you will encounter certain issues that will cause your pull request to get held up by test failures. This article describes a workaround for a similar issue I encountered when attempting to have my modifications approved and merged into the master branch.
```

## Why is this needed?
 

> While contributing to [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes) for some minor documentation changes, the pushed changes needed to be updated with other verified contents of the entire documentation. So, in order for the change to take effect, a single command must be performed to ensure that all tests on the CI pipeline pass. The single command `make update` runs all presubmission verification tests. For some reason on the windows `wsl` environment the tests, specifically the [update-openapi-spec.sh](https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-openapi-spec.sh) script, failed (in my case, take a look at the conversation [here](https://github.com/kubernetes/kubernetes/pull/107691)), eventually failing the `pull-kubernetes-verify` tests.

Reviewer 2: The link `https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-openapi-spec.sh` should be `https://github.com/kubernetes/kubernetes/blob/master/hack/update-openapi-spec.sh`

(it's OK to replace `master` with a specific commit).

Reviewer 2: Change “windows `wsl`” to “WSL”, or write “Windows Subsystem for Linux”

### System info:

| OS name | Version | System type |Memory| WSL |
| ------- | ------- | ----------- | ---  | --- |

> |Windows 11 home|10.0.22000 Build 22000|x64-based|8.00 GB|Ubuntu-20.04 (v2)|

Reviewer 2: Is this needed? I'd omit it.

Author: I agree, it doesn't fit entirely to the article's context.

You might encounter the following on your PR

The tests failing the particular issue:

![Failed Test Cases](failing_test_cases.png)

Consecutively, 

![Failed Test Cases robot comment](k8s_ci_failed_tests.png)

Additionally one can check the failed test via the `link` provided under details in the above image.

### Potential solution
Run the failing `.sh` scripts individually known from the CI job output, to generate the expected files to fix up the failures. The `.sh` scripts can be found residing under the `hack/` directory at the root of the `kubernetes/kubernetes` code base. 

>  `kubernetes/kubernetes &rarr; hack/ &rarr; .sh update files`

Reviewer 2: ```suggestion
`kubernetes/kubernetes` &rarr; `hack/*.sh`
```
or
```suggestion
`hack/*.sh`
```
or
```suggestion
See the [`hack/`](https://github.com/kubernetes/kubernetes/tree/master/hack) directory and look for shell scripts (`*.sh`).
```

In this particular case, these files were to be run:

* [hack/update-generated-protobuf.sh](https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-generated-protobuf.sh) 
* [hack/update-generated-swagger-docs.sh](https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-generated-swagger-docs.sh)

> * [hack/update-openapi-spec.sh](https://github.com/SubhasmitaSw/kubernetes/blob/master/hack/update-openapi-spec.sh)

Reviewer 2: These links are wrong. Make sure to reference https://github.com/kubernetes/kubernetes/

Author: Done!

After you run these, you might see either of the below issues:

---
Using Codespaces:

![codespaces error output](codespaces.png)
![codespaces error output](codespaces2.png)

---

Using Visual Studio Code:

![vscode error output](vscode.png)

### Potential solution to the above error:
1. Remove the Makefile and Makefile.generated_files files:

```
rm Makefile Makefile.generated_files

> ```

Reviewer 2: Indent this text by 3 spaces.

2. Create the symlinks:

```
ln -s build/root/Makefile Makefile
```

```
ln -s build/root/Makefile.generated_files Makefile.generated_files

> ```

Reviewer 2: Indent this text by 3 spaces.

> > PS: takes adequate time to generate the scripts on windows.

Reviewer 2: ```suggestion
(this can take a while)
```
(is that what you meant to say)

> ## The current situation

Reviewer 2: ```suggestion
## The problem
```

> Many contributors do not have access to powerful environments in which to run `make update` or `make verify`. They can utilise *vscode/wsl/codespaces* and other tools to recommend modifications, but they might get tripped up by `make verify` scripts because in many cases, we end up with files that needs to be re-generated. It's a tall order for them to scan the build log from `make verify` to determine which specific scripts in `hack/` directory they need to run.

Reviewer 2: How about:
```suggestion
There are lots of folks who'd like to make a small contribution to Kubernetes, but
find that setting up a development environment can be more effort that the code change
itself.
Many contributors do not have access to powerful environments in which to run
`make update` or `make verify`. Even people who can use tools such as
_GitHub CodeSpaces_, _Multipass_, _Vagrant_, _Visual Studio Code_, or
_Windows Subsystem for Linux_ can still get tripped up by failing the `make verify`
script that runs as part of CI testing.

You might be a contributor who has tried this and, like me, ended up with files that needs to be
re-generated. It can be a tall order to go from scanning the `make verify` build log
to determining which specific script(s) in `hack/` directory you need to run.
```

I listed some more tools; I didn't think it felt fair to only mention Microsoft ones.

### Solution

> Two solutions are proposed by [@dims](https://github.com/dims) for the time being:

Reviewer 2: Is this accurate
```suggestion
When Kubernetes contributor [@dims](https://github.com/dims) hears about my challenges,
he proposed two solutions, available in different timescales:
```

* **Short Term:** add a CI job that folks can trigger when they need it, that runs `make update` and provides a zip archive that they can be told to download. The archive includes all the files that changed as a result of running `make update`.

> *  **Long Term:** add a new Prow bot command that generates an additional commit on their PR.

Reviewer 2: You could make the text “Short Term” and “Long Term” be hyperlinks to the relevant headings.

> ### Solution
> 
> Two solutions are proposed by [@dims](https://github.com/dims) for the time being:
> 
> * **Short Term:** add a CI job that folks can trigger when they need it, that runs `make update` and provides a zip archive that they can be told to download. The archive includes all the files that changed as a result of running `make update`.
> *  **Long Term:** add a new Prow bot command that generates an additional commit on their PR.
> 
> ### Short Term Solution

Reviewer: So, here's my suggestion for easier-to-follow text for this:

```suggestion
### Solutions

According to [@dims](https://github.com/dims), a long term solution for this problem would be to add a new Prow bot command that generates an additional commit on their PR.

In the meantime, the short term workaround is to add a CI job that folks can trigger when they need it, that runs `make update` and provides a zip archive that they can be told to download. The archive includes all the files that changed as a result of running `make update`.

### Implementing the Short Term Workaround
```

This avoids readers getting lost in the text flow.

Also -- is there an issue for the Prow bot?  If so, we should link to it in this text.

Author: We don't currently have an issue created for the Prow bot.
How about referencing this [issue](https://github.com/kubernetes/kubernetes/issues/107726) where it was mentioned?

Reviewer: I was asking if there was an issue that you could reference.  Someone should maybe create one, if that's our long-term solution

Author: I'll do it.

Reviewer: that might be up to DIms though.  So ... no link.

Author: I've created an [issue](https://github.com/kubernetes/kubernetes/issues/109374#issuecomment-1092155063) in this regard, let me know if this is fine to include.

Reviewer: Yes, please include it.

> Once we realized the problem was that some of the verify scripts needed either linux to work better or a large CPU/Memory even with linux, we decided to add a new CI job. This CI job is named “pull-kubernetes-update”. You can trigger this CI job by running “/test pull-kubernetes-update” bot command in any PR. This CI job runs `make update` and generates a zip file named “updated-files.zip” in the artifacts directory for that job. Folks can then download the zip which has the changes made when `make update` was run and update their PR with the newly updated code.

Reviewer 2: Repeating an earlier suggestion
```suggestion
The problem was that for some of the verify scripts to work well, the scripts needed both Linux
and plenty of local resources (CPU, memory). Once we realized this issue, we decided to add a
new CI job. This CI job is named `pull-kubernetes-update`. You can trigger this CI job by commenting
`/test pull-kubernetes-update` as a Prow bot command in any Kubernetes PR. This CI job runs
`make update` and then generates a zip archive named `updated-files.zip` in the artifacts directory for
that job.
You can then download the archive which has the changes made when `make update` was run;
once the download is finished, you can update your PR with the newly updated code.
```

> **Steps:**

Reviewer 2: ```suggestion
#### Using the short-term solution

Here's what you can do as a contributor to get automated help updating your pull request with generated files.
```

> Make sure to rebase your 'working branch' to 'master' to align with current changes made to the repository. 

Reviewer 2: ```suggestion
Make sure to rebase your working branch against the latest base branch (usually this is `master`), so
that your PR includes the most recent upstream commits. Remember to push or force-push your changes.
```

On your PR, write a new comment that contains only the text: 

```
/test pull-kubernetes-update
```

![pull-kubernetes-update command](tpku_command.png)

You will see the automated check lists and details.

![command preview on github](checklists.png)

Once the checks are complete, click the `Details` link on `pull-kubernetes-update` job to go to the artifacts directory and  download the `updated-files.zip` file.

![automated checklist of tests appearing after update command execution](completed_checklist.png)

![completed checklist of tests](artifacts.png)

![artifacts directory](jenkins.png)

Now, update the PR by adding the extracted files you downloaded. 

> 

Reviewer 2: There's no heading or steps about using the long term solution. Folks might wonder where that has gone.

My guess is that the long term solution is still a work in progress and you can't use it yet? Whatever the story is, let's make that really clear.

Reviewer: I have an idea how to arrange this better, suggestion coming.

---
layout: blog
title: >-

>     "K8s Triage bot Helper CI Job: ‘*make update*’"

Reviewer 2: ```suggestion
    Kubernetes Build Helper CI job: Run "make update" within CI
```

Author: I was changing those, I accidentally deleted the last changes when my pc went down. I shall push them soon.

## Why is this needed?
 
While contributing to [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes) for some minor documentation changes, the pushed changes needed to be updated with other verified contents of the entire documentation. So, in order for the change to take effect, a single command must be performed to ensure that all tests on the CI pipeline pass. The single command `make update` runs all presubmission verification tests. For some reason on the "Windows Subsystem for Linux" environment the tests, specifically the [update-openapi-spec.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/update-openapi-spec.sh) script, failed (in my case, take a look at the conversation [here](https://github.com/kubernetes/kubernetes/pull/107691)), eventually failing the `pull-kubernetes-verify` tests.

You might encounter the following on your PR

The tests failing the particular issue:

![Failed Test Cases](failing_test_cases.png)

Consecutively, 

> ![Failed Test Cases robot comment](k8s_ci_failed_tests.png)

Author: I'm concerned about this too, what could be causing this? Maybe the file is moved to some location not containing the images
